Mathlib/SetTheory/Ordinal/Arithmetic.lean

```diff
@@ -372,9 +313,99 @@ theorem mk_Iio_ordinal (o : Ordinal.{u}) :
 theorem mk_initialSeg (o : Ordinal.{u}) :
     #{ o' : Ordinal | o' < o } = Cardinal.lift.{u + 1} o.card := mk_Iio_ordinal o
 
+/-! ### The predecessor of an ordinal -/
 
-/-! ### Normal ordinal functions -/
+/-- The ordinal predecessor of `o` is `o'` if `o = succ o'`, and `o` otherwise. -/
+def pred (o : Ordinal) : Ordinal :=
+  isSuccPrelimitRecOn o (fun a _ ↦ a) (fun a _ ↦ a)
+
+@[simp]
+theorem pred_succ (o) : pred (succ o) = o :=
+  isSuccPrelimitRecOn_succ ..
+
+theorem pred_eq_of_isSuccPrelimit {o} : IsSuccPrelimit o → pred o = o :=
+  isSuccPrelimitRecOn_of_isSuccPrelimit _ _
 
+alias _root_.Order.IsSuccPrelimit.pred_eq := pred_eq_of_isSuccPrelimit
+
+theorem _root_.Order.IsSuccLimit.pred_eq {o} (ho : IsSuccLimit o) : pred o = o :=
+  ho.isSuccPrelimit.pred_eq
+
+@[simp]
+theorem pred_zero : pred 0 = 0 :=
+  pred_eq_of_isSuccPrelimit isSuccPrelimit_zero
+
+theorem pred_le_self (o) : pred o ≤ o := by
+  obtain ⟨a, rfl⟩ | ho := mem_range_succ_or_isSuccPrelimit o
+  · simpa using le_succ a
+  · rw [pred_eq_of_isSuccPrelimit ho]
+
+theorem self_le_succ_pred (o) : o ≤ succ (pred o) := by
+  obtain ⟨a, rfl⟩ | ho := mem_range_succ_or_isSuccPrelimit o
+  · simp
+  · simpa [ho.pred_eq] using le_succ o
```

Comment:
Can you prove that `pred` and `succ` form a galois connection? namely `pred a <= b iff a <= succ b` and `pred (succ a) = a`

Reply:
Both of these theorems exist, see `pred_succ_gc` and `pred_succ`.